storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,670 @@
+        add_operations_[attr_typeid]
+          = BinaryOperationFactory::GetBinaryOperation(BinaryOperationID::kAdd)
+              .makeUncheckedBinaryOperatorForTypes(TypeFactory::GetType(attr_typeid),
+                                                   TypeFactory::GetType(attr_sum_typeid));
+      }
+    }
+
+    // Initialize comparison map. Maps attribute type to comparison function.
+    int attr_typeid = static_cast<int>(attribute_type.getTypeID());
+    if (less_comparisons_[attr_typeid] == nullptr) {
+      DCHECK(equal_comparisons_[attr_typeid] == nullptr);
+      less_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kLess)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+      equal_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kEqual)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+    }
+
+    // Map the attribute's id to its entry's index.
+    attribute_to_entry_[attribute] = indexed_attribute_num;
+
+    if (!header_->index_consistent) {
+      resetEntry(entry, attribute, attribute_type);
+    } else {
+      // If the data held by min/max is out of line, we must retrieve it as we initialize.
+      if (!TypedValue::RepresentedInline(entry->type_id)) {
+        // First, set to 0 so that we don't try to free invalid memory on copy.
+        // Next, copy from the tuple store. This will copy and give us ownership of out of line data.
+        if (entry->min_entry_ref.valid) {
+          new (&entry->min_entry_ref.value) TypedValue();
+          entry->min_entry_ref.value = tuple_store
+              .getAttributeValueTyped(entry->min_entry_ref.entry_ref_tuple, entry->attribute);
+          entry->min_entry_ref.value.ensureNotReference();
+        }
+        if (entry->max_entry_ref.valid) {
+          new (&entry->max_entry_ref.value) TypedValue();
+          entry->max_entry_ref.value = tuple_store
+              .getAttributeValueTyped(entry->max_entry_ref.entry_ref_tuple, entry->attribute);
+          entry->max_entry_ref.value.ensureNotReference();
+        }
+      }
+    }
+  }
+  initialized_ = true;
+}
+
+SMAIndexSubBlock::~SMAIndexSubBlock() {
+  // Any typed values which have out of line data must be cleared.
+  freeOutOfLineData();
+
+  // Clear any operators which have been allocated.
+  for (int i = 0; i < kNumTypeIDs; ++i) {
+    if (add_operations_[i] != nullptr) {
+      delete add_operations_[i];
+    }
+    if (equal_comparisons_[i] != nullptr) {
+      delete equal_comparisons_[i];
+    }
+    if (less_comparisons_[i] != nullptr) {
+      delete less_comparisons_[i];
+    }
+  }
+
+  // Delete the operator arrays.
+  delete[] add_operations_;
+  delete[] equal_comparisons_;
+  delete[] less_comparisons_;
+  delete[] attribute_to_entry_;
+}
+
+void SMAIndexSubBlock::resetEntry(SMAEntry *entry,
+                                  attribute_id attribute,
```

Comment:
Can we have a more informational name like ``attr_id`` instead of ``attribute``?